registry/storage/tagstore.go

```diff
@@ -150,96 +149,84 @@ func (ts *tagStore) linkedBlobStore(ctx context.Context, tag string) *linkedBlob
 		// This tag store has been initialized but not yet populated
 		break
-	case nil:
-		break
 	default:
 		return nil, err
 	}
 
-	limiter := make(chan struct{}, ts.lookupConcurrencyFactor)
-	errChan := make(chan error, len(allTags))
-	tagChan := make(chan string, len(allTags))
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	type lookupResult struct {
+	    tag string
+	    error error
+        }
 
-	var tags []string
-	wg := sync.WaitGroup{}
-	wg.Add(len(allTags))
-	for _, tag := range allTags {
-		limiter <- struct{}{}
-		go func(tag string) {
-			defer wg.Done()
-			defer func() {
-				<-limiter
-			}()
-			tagLinkPathSpec := manifestTagCurrentPathSpec{
-				name: ts.repository.Named().Name(),
-				tag:  tag,
-			}
+	inputChan := make(chan string)
+	outputChan := make(chan lookupResult, len(allTags))
 
-			tagLinkPath, _ := pathFor(tagLinkPathSpec)
-			// If context is done stop short as the parent has exited with an error
-			if ctx.Err() != nil {
-				return
-			}
-			tagDigest, err := ts.blobStore.readlink(ctx, tagLinkPath)
-			if err != nil {
-				switch err.(type) {
-				case storagedriver.PathNotFoundError:
-					return
+	workersWaitGroup := sync.WaitGroup{}
+	workersWaitGroup.Add(ts.lookupConcurrencyFactor)
+
+	for i := 0; i < ts.lookupConcurrencyFactor; i++ {
+		go func() {
+			defer workersWaitGroup.Done()
+			for tag := range inputChan {
+				ctxErr := ctx.Err()
+				if ctxErr != nil {
+					outputChan <- lookupResult{tag, ctxErr}
 				}
-				select {
-				case <-ctx.Done():
-					return
-				default:
+
+				tagLinkPathSpec := manifestTagCurrentPathSpec{
+					name: ts.repository.Named().Name(),
+					tag:  tag,
 				}
-				errChan <- err
-				return
-			}
 
-			if tagDigest == desc.Digest {
+				tagLinkPath, _ := pathFor(tagLinkPathSpec)
+				tagDigest, err := ts.blobStore.readlink(ctx, tagLinkPath)
 
-				select {
-				case <-ctx.Done():
-					return
-				default:
+				if err != nil {
+					switch err.(type) {
+					case storagedriver.PathNotFoundError:
+						continue
+					}
+					outputChan <- lookupResult{tag, err}
+				}
+
+				if tagDigest == desc.Digest {
+					outputChan <- lookupResult{tag, err}
 				}
-				tagChan <- tag
-				return
 			}
-		}(tag)
+		}()
+        }
+
+	for _, tag := range allTags {
+		inputChan <- tag
 	}
+	close(inputChan)
 
-	go func() {
-		wg.Wait()
-		close(tagChan)
-
-	}()
-read:
-	for {
-		select {
-		case err, workRemaining := <-errChan:
-			if !workRemaining {
-				break read
-			}
-			cancel()
-			return nil, err
-		case tag, workRemaining := <-tagChan:
-			if !workRemaining {
-				close(errChan)
-				break read
-			}
-			tags = append(tags, tag)
-		}
+	workersWaitGroup.Wait()
 
+	close(outputChan)
+
+	var tags []string
+	var error error
+
+	for lookupResult := range outputChan {
+              if lookupResult.error != nil {
+		      error = lookupResult.error
+		      break
+	      }
+	      tags = append(tags, lookupResult.tag)
 	}
 
-	return tags, nil
+	return tags, error
 }
```

Comment:
```go
type atomicError struct {
	mu  sync.Mutex
	err error
}

func (e *atomicError) Store(err error) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.err = err
}

func (e *atomicError) Load() error {
	e.mu.Lock()
	defer e.mu.Unlock()
	return e.err
}

// Lookup recovers a list of tags which refer to this digest.  When a manifest is deleted by
// digest, tag entries which point to it need to be recovered to avoid dangling tags.
func (ts *tagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
	allTags, err := ts.All(ctx)
	switch err.(type) {
	case nil:
		break
	case distribution.ErrRepositoryUnknown:
		// This tag store has been initialized but not yet populated
		break
	default:
		return nil, err
	}

	lookupErr := &atomicError{}

	inputChan := make(chan string)
	outputChan := make(chan string, len(allTags))

	workersWaitGroup := sync.WaitGroup{}
	workersWaitGroup.Add(ts.lookupConcurrencyFactor)

	for i := 0; i < ts.lookupConcurrencyFactor; i++ {
		go func() {
			defer workersWaitGroup.Done()
			for tag := range inputChan {
				tagLinkPathSpec := manifestTagCurrentPathSpec{
					name: ts.repository.Named().Name(),
					tag:  tag,
				}

				tagLinkPath, _ := pathFor(tagLinkPathSpec)
				tagDigest, err := ts.blobStore.readlink(ctx, tagLinkPath)

				if err != nil {
					switch err.(type) {
					case storagedriver.PathNotFoundError:
						continue
					}
					lookupErr.Store(err)
				}

				if tagDigest == desc.Digest {
					outputChan <- tag
				}
			}
		}()
	}

	for _, tag := range allTags {
		if lookupErr.Load() != nil {
			break
		}

		select {
		case <-ctx.Done():
			break
		default:
		}

		select {
		case <-ctx.Done():
			break
		case inputChan <- tag:
		}
	}
	close(inputChan)

	workersWaitGroup.Wait()

	close(outputChan)

	if err := lookupErr.Load(); err != nil {
		return nil, err
	}

	tags := make([]string, 0, len(outputChan))
	for tag := range outputChan {
		tags = append(tags, tag)
	}

	return tags, nil
}
```